tools/lkl/lib/virtio_net.c

```diff
@@ -238,3 +238,42 @@ int lkl_netdev_add(struct lkl_netdev *nd, void *mac)
 
 	return ret;
 }
+
+/* Return 0 for success, 1 for failure. */
+static int lkl_netdev_remove(struct virtio_net_dev *dev)
+{
+	if (!dev->nd->ops->close)
+		/* Can't kill the poll threads, so we can't do
+		 * anything safely. */
+		return 1;
+
+	if (dev->nd->ops->close(dev->nd) < 0)
+		/* Something went wrong */
+		return 1;
+
+	virtio_dev_cleanup(&dev->dev);
+
+	lkl_host_ops.mem_free(dev->nd);
+	free_queue_locks(dev->queue_locks, NUM_QUEUES);
+	lkl_host_ops.mem_free(dev);
+
+	return 0;
+}
+
+int lkl_netdevs_remove(void)
+{
+	int i = 0, failure_count = 0;
+
+	for (; i < registered_dev_idx; i++)
+		failure_count += lkl_netdev_remove(registered_devs[i]);
+
+	if (failure_count) {
+		lkl_printf("WARN: failed to free %d of %d netdevs.\n",
+			failure_count, registered_dev_idx);
+		return -1;
+	} else {
+		lkl_printf("Freed all %d netdevs successfully.\n",
```

Comment:
I prefer no prints in case of success. In the test below we can check for failures instead.